sdk/metric/provider.go

```diff
@@ -68,6 +71,10 @@ func NewMeterProvider(options ...Option) *MeterProvider {
 //
 // This method is safe to call concurrently.
 func (mp *MeterProvider) Meter(name string, options ...metric.MeterOption) metric.Meter {
+	if mp.stopped.Load() {
```

Comment:
Bit of a synchronization nit, but it's possible for this atomic boolean to return true, then for a `Shutdown()` to be called and complete successfully, then for this method to return a valid (not noop) Meter.

Another option would be to use `sync.RWMutex`, get an `RLock` (a read lock, of which there can be multiple simultaneously) here and and a regular `Lock` in the Shutdown method -- something like:

```go
type MeterProvider struct {
    sync.RWMutex
    stopped bool
}

func (mp *MeterProvider) Meter(name string, options ...metric.MeterOption) metric.Meter {
    mp.RLock()
    defer mp.RUnlock()

    if mp.stopped {
        return noop.Meter{}
    }
}

func (mp *MeterProvider) Shutdown(ctx context.Context) error {
    mp.Lock()
    defer mp.Unlock()

    mp.stopped = true
}
```

Reply:
I do not believe it is needed.


From https://go.dev/ref/mem#atomic:

> The APIs in the [sync/atomic](https://go.dev/pkg/sync/atomic/) package are collectively “atomic operations” that can be used to synchronize the execution of different goroutines. If the effect of an atomic operation A is observed by atomic operation B, then A is synchronized before B.
>
> All the atomic operations executed in a program behave as though executed in some sequentially consistent order.

From https://pkg.go.dev/sync/atomic:

> Additionally, all the atomic operations executed in a program behave as though executed in some sequentially consistent order. This definition provides the same semantics as C++'s sequentially consistent atomics and Java's volatile variables.
>
> The preceding definition has the same semantics as C++’s sequentially consistent atomics and Java’s volatile variables.

From: https://go101.org/article/memory-model.html:

> Since Go 1.19, the Go 1 memory model documentation formally specifies that all atomic operations executed in Go programs behave as though executed in some sequentially consistent order. 

Therefore, the code after `mp.stopped.Store(true)` won't be reordered (executed) before it. As well as the code before `mp.stopped.Load()` won't be reordered (executed) after it.

Reply:
Thank you @pellared for the research! 🌷 
I have added a comment about that so folks reading the code later don't trip up on this too.

Reply:
One more reference: https://en.cppreference.com/w/cpp/atomic/memory_order#Sequentially-consistent_ordering

Reply:
Should be addressed via https://github.com/open-telemetry/opentelemetry-go/pull/4154/commits/19eff3a68651b1c20fd44c3681658e639c244afd

@pmcollins please double-check

Reply:
Here you are correct. However, I cannot imagine it could cause any bug. The meters which were returned before `Shutdown` should still "be working". As far as I understand returning no-op meter is just an optimization.

Reply:
Yeah, it's hard to see this causing a problem unless the caller is relying on the returned Meter being a noop to determine if the Provider has finished shutting down 🤷. That said, it could be worth addressing just for that scenario, and for enhancing overall correctness. But I'll defer to your judgement and that of the author here.

Reply:
I think using a lock here is going to produce more consistent results and the contention introduced would be acceptable given the infrequency `Shutdown` and `Meter` are called. That said, I don't think it is a critically needed change.

@dmathieu thoughts? I'm fine if you want to leave it as is or change it.

Reply:
Either works for me. While I don't think using a mutex is really useful, the complexity between one and using an atomic boolean is also not very different.